File: 2015/23.py
from Read import *
import re
import math
import itertools
import functools

# Part 1

inp = rdl("23.txt")

a = 1
b = 0

i = 0
while i < len(inp):
    line = inp[i]
    ss = line.split(' ')
    instr = ss[0]

    if instr == 'hlf':
        if ss[1] == 'a':
            a /= 2
        elif ss[1] == 'b':
            b /= 2
        i += 1
    elif instr == 'tpl':
        if ss[1] == 'a':
            a *= 3
        elif ss[1] == 'b':
            b *= 3
        i += 1
    elif instr == 'inc':
        if ss[1] == 'a':
            a += 1
        elif ss[1] == 'b':
            b += 1
        i += 1
    elif instr == 'jmp':
        amt = int(ss[1])
        i += amt
    elif instr == 'jie':
        reg = ss[1][0]
        amt = int(ss[2])

        if reg == 'a' and a % 2 == 0:
            i += amt
        elif reg == 'b' and b % 2 == 0:
            i += amt
        else:
            i += 1
    elif instr == 'jio':
        reg = ss[1][0]
        amt = int(ss[2])

        if reg == 'a' and a == 1:
            i += amt
        elif reg == 'b' and b == 1:
            i += amt
        else:
            i += 1

print(b)

# 6:24 (3rd)

# Part 2

# 6:37 (3rd)

File: v2/2015/23.R

# Load --------------------------------------------------------------------

rm(list = ls())
source("utilities.R")

# Problem -----------------------------------------------------------------

# inp <- strsplit("inc a
# jio a, +2
# tpl a
# inc a", "\n")[[1]]  # Example

inp <- get_input("2015/23", parse = F, user = "blakeebets", cache = T)

reg <- c('a' = 1, 'b' = 0)
i <- 1

while(TRUE) {
    if(i < 1 || i > length(inp)) {
        break
    }
    
    line <- inp[i]
    ss <- strsplit(line, " ")[[1]]
    instr <- ss[1]
    
    if(instr == "hlf") {
        reg[[ss[2]]] <- reg[[ss[2]]] / 2
        i <- i + 1
    } else if(instr == "tpl") {
        reg[[ss[2]]] <- reg[[ss[2]]] * 3
        i <- i + 1
    } else if(instr == "inc") {
        reg[[ss[2]]] <- reg[[ss[2]]] + 1
        i <- i + 1
    } else if(instr == "jmp") {
        amt <- as.numeric(ss[2])
        i <- i + amt
    } else if(instr == "jie") {
        target <- gsub(",", "", ss[2])
        
        if(reg[[target]] %% 2 == 0) {
            amt <- as.numeric(ss[3])
            i <- i + amt
        } else {
            i <- i + 1
        }
    } else if(instr == "jio") {
        target <- gsub(",", "", ss[2])
        
        if(reg[[target]] == 1) {
            amt <- as.numeric(ss[3])
            i <- i + amt
        } else {
            i <- i + 1
        }
    }
}

reg

# 5:58 (2nd)

# 6:08 (1st)



